### PR TITLE
58 Allow adding maps to widgets

### LINKED ISSUE #58: Allow adding maps to widgets
- [x] add google maps plugin 
   - [google_maps_flutter | Flutter Package](https://pub.dev/packages/google_maps_flutter) 
- [x] add permissions handler plugin 
   - [permission_handler | Flutter Package](https://pub.dev/packages/permission_handler) 
- [x] check location permission on app load and request if status not determine 

#### Android 
- [x] add a valid API Key to local.properties 
- [x] access the `local.properties` file in app/build.gradle to get API Key 
- [x] use the API key in the application manifest 
   - see [Inject build variables into the manifest  |  Android Developers](https://developer.android.com/studio/build/manifest-build-variables) 

#### iOS 
- [x] add `Flutter/MapsKey.xcconfig` with `MAPS_API_KEY=...`
- [x] add to Info.plist 
```sh
<key>maps_api_key</key>
<string>$(MAPS_API_KEY)</string>
<key>io.flutter.embedded_views_preview</key>
<true/>
```
- [x] add to `AppDelegate.swift` to retrieve the API key and provide it to the Maps
```Swift
let mapsKey = Bundle.main.infoDictionary?["maps_api_key"] as! String
GMSServices.provideAPIKey(mapsKey)
```

### PR DESCRIPTION
**Pull request by nickmeinhold**
fixes #58

### CODE REVIEW

**Comment by nickmeinhold** [comment]
@tamari-gray we need an API key: 

https://pub.dev/packages/google_maps_flutter#getting-started 

Are you happy to create one and email it to me or something?  Thanks!

**Comment by nickmeinhold** [comment]
> Are you happy to create one and email it to me or something? Thanks!

Actually I can just use my own, no need to go sharing yours :-)  Ignore me!

**Comment by nickmeinhold** [comment]
Ugh I forgot how annoying this part is (avoiding the API key being in version control). I'll figure it though.  

In case my previous ramblings have confused the issue, you'll need to:

-   Get an API key at <https://cloud.google.com/maps-platform/>.
    - (so yeah no need to share it with me, but it will need to be the one that gets used when the app is deployed, presumably by CD in GitHub actions) 

-   Enable Google Map SDK for each platform.

    -   Go to [Google Developers Console](https://console.cloud.google.com/).
    -   Choose the project that you want to enable Google Maps on.
    -   Select the navigation menu and then select "Google Maps".
    -   Select "APIs" under the Google Maps menu.
    -   To enable Google Maps for Android, select "Maps SDK for Android" in the "Additional APIs" section, then select "ENABLE".
    -   To enable Google Maps for iOS, select "Maps SDK for iOS" in the "Additional APIs" section, then select "ENABLE".
    -   Make sure the APIs you enabled are under the "Enabled APIs" section.
-   You can also find detailed steps to get start with Google Maps Platform [here](https://developers.google.com/maps/gmp-get-started).

OK I'll stop now! :-)

**Comment by tamari-gray** [comment]
Omg thanks so much Nick :pray:

**Comment by nickmeinhold** [comment]
OK so we could manage getting the API key with dart-defines: 

[Flutter 1.17 — no more Flavors, no more iOS Schemas. Command argument that changes everything | by Denis Beketsky | ITNEXT](https://itnext.io/flutter-1-17-no-more-flavors-no-more-ios-schemas-command-argument-that-solves-everything-8b145ed4285d) 

It's a pretty cool new(ish) feature of dart/flutter -> being able to provide values as command line arguments at compile time that can be accessed by native code at run time.  So many possibilities :-)  

There's a fair bit of plumbing involved though and it would be easier to just use the last half of the setup described in that article (without the dart-define part, just having the API key in files that are not checked in).  dart-defines, being new and not an official part of the language, could potentially change too (although I think it's unlikely). 

I'm going to have a think about it and maybe we can chat about it. Have a great day!

**Comment by nickmeinhold** [comment]
After thinking on it a little - the overall task of doing this can be split into a few parts without any overlap: 
1. using `dart-define` to get values supplied at compile time into variables at run time 
1. adding native code that retrieves the dart-define values 

So I'm thinking we can do the second part and leave the first part for another ticket (if at all) - I'll create an issue for discussion.

**Comment by nickmeinhold** [comment]
@tamari-gray when you feel like it the "Add a map widget and test" commit on this branch can act as a test for the Maps API key you're gonna get... if. you add `maps.API_KEY=... ` to `android/local.properties` you should get a map shown on app load. 👍

**Comment by nickmeinhold** [comment]
@tamari-gray I know you're likely not around today (good job on having a rest day!) so no expectation of when you'll get to this... have an awesome rest mate!

File: android/app/build.gradle [hunk 1]
```diff
@@ -21,13 +21,18 @@ if (flutterVersionName == null) {
     flutterVersionName = '1.0'
 }
 
+def mapsKeyProperty = localProperties.getProperty('maps.API_KEY')
+if (mapsKeyProperty == null) {
+    throw new GradleException("Google Maps API Key not found. Define API key with maps.API_KEY in the local.properties file.")
```
Comment (tamari-gray) [inline android/app/build.gradle:26]: love it, so clear

**Review by tamari-gray — APPROVED** [review]

**Comment by tamari-gray** [comment]
thanks for the beautiful work @nickmeinhold 🙏

**Pull request merged**